echopype/calibrate/calibrate_base.py

```diff
@@ -7,13 +13,158 @@
+            )
+            extrap_unique_idx = {
+                dim: np.unique(extrap[dim], return_index=True)[1] for dim in dims
+            }
+            extrap = extrap.isel(**extrap_unique_idx)
+            interp = env_params.interp(
+                {dim: echodata.platform[dim].data for dim in dims},
+                method=self.interp_method,
+            )
+            interp_unique_idx = {
+                dim: np.unique(interp[dim], return_index=True)[1] for dim in dims
+            }
+            interp = interp.isel(**interp_unique_idx)
+
+            if self.extrap_method is not None:
+                less = extrap.sel(
+                    {
+                        dim: extrap[dim][extrap[dim] < min_max[dim]["min"]]
+                        for dim in dims
+                    }
+                )
+                middle = interp.sel(
+                    {
+                        dim: interp[dim][
+                            np.logical_and(
+                                interp[dim] >= min_max[dim]["min"],
+                                interp[dim] <= min_max[dim]["max"],
+                            )
+                        ]
+                        for dim in dims
+                    }
+                )
+                greater = extrap.sel(
+                    {
+                        dim: extrap[dim][extrap[dim] > min_max[dim]["max"]]
+                        for dim in dims
+                    }
+                )
+
+                # remove zero length dims (xarray does not support in combine_by_coords)
+                non_zero_dims = [
+                    ds
+                    for ds in (less, middle, greater)
+                    if all(dim_len > 0 for dim_len in ds.dims.values())
+                ]
+                env_params = xr.combine_by_coords(non_zero_dims)
+
+        # if self.data_kind == "organized":
+        #     # get platform latitude and longitude indexed by ping_time
+        #     interp_plat = echodata.platform.interp(
+        #         {"time": echodata.platform["ping_time"]}
+        #     )
+        #     # get env_params latitude and longitude indexed by ping_time
+        #     env_params = env_params.interp(
+        #         {
+        #             "latitude": interp_plat["latitude"],
+        #             "longitude": interp_plat["longitude"],
+        #         }
+        #     )
+
+        return {var: env_params[var] for var in ("temperature", "salinity", "pressure")}
+
 
 class CalibrateBase(abc.ABC):
     """Class to handle calibration for all sonar models."""
 
-    def __init__(self, echodata):
+    def __init__(self, echodata, env_params=None):
         self.echodata = echodata
-        self.env_params = None  # env_params are set in child class
+        if isinstance(env_params, EnvParams):
+            env_params = env_params._apply(echodata)
+        elif env_params is None:
+            env_params = {}
```

Comment:
Missing an `else` case to catch other possible inputs?

Reply:
I think this is just for converting `env_params` into a dictionary format. I believe the accepted values are `EnvParams`, `None`, or a dictionary, so if it isn't `EnvParams` or `None` then it should be a dictionary which is OK

Reply:
What if people put in `env_params=""` ?

Reply:
Is that something they can do? I didn’t think that was allowed

Reply:
I was thinking that people won't necessary read the docstring, so it would be useful to a thing to catch those and print out an error message. Maybe I missed something?